WIP: first careful tests to get Debian working as Salt client

rpasche opened the pull request:
These are the first modifications I have done (with further
modifications on the salt client side) to get a Debian
system registered as a Salt client.

## What does this PR change?

This is PR should track #261 I created to somehow track the issues and problems I had while trying to register a Debian system as "Salt" client into Uyuni. It is not yet really meant to be a *real* PR to be merged into Uyuni 

## GUI diff

No difference.

Before:

After:

- [X] **DONE**

## Documentation
- No documentation needed: Still testing 

- [X] **DONE**

## Test coverage
- No tests: Still testing.

- [X] **DONE**

## Links

Fixes #
Tracks # #261 

- [X] **DONE**

juliogonzalez commented:
So far I think it looks good!

Maybe @mcalmer, @mateiw or @meaksh can have a look now, or when the PR is complete.

rpasche commented:
I need some more help on this. I'm pretty new to Salt and have a major problem right now. As I mentioned in issue #261 , I modified the clients ```pkg.info_installed``` function to return the attributes, that get requested within ```packages/profileupdate```. But these attributes (```epoch, arch, release, install_date_time_t```) do not exist on Debian systems. So I would like to call that function with the available attributes for Debian and (e.g. ```architecture, version, install_date```, then transform these and return that to the salt-master. Also, the Debian ```version``` still includes the ```epoch``` and ```release``` (if present.
How should this be done?

This is an example, after I modified ```pkg.info_installed``` to return the Debian attributes for a package.
```
root@uyuni-test-client:~# salt-call -l error pkg.info_installed attr='architecture,version,install_date' attr
local:
    ----------
    attr:
        ----------
        architecture:
            amd64
        install_date:
            2018-10-24T08:17:51Z
        version:
            1:2.4.47-2+b2
root@uyuni-test-client:~#
```
As you can see, we then need to transform the ```version``` (to split into ```epoch```, ```version``` and ```release```). Also we need to add the attribute ```install_date_time_t```
Any help will be appreciated.

rpasche commented:
Hmm...I thought I had an solution. But this does not work.

I used the module ```sumautil.py``` and added a custom method ```debian_info_installed``` which is basically a copy of ```aptpkg/info_installed```. If I call the function manually, everything works, but the package refresh within webui fails. Looks like this is because of the missing serialization/deserialization within ```java/code/src/com/suse/manager/webui/utils/salt/custom/PkgProfileUpdateSlsResult.java```

See the difference

```
uyuni-test:/usr/share/susemanager/salt/packages # salt 'uyuni-test-client.domain' -l error pkg.info_installed attr
uyuni-test-client.swr.ard:
    ----------
    attr:
        ----------
        architecture:
            amd64
        description:
            Utilities for manipulating filesystem extended attributes
             A set of tools for manipulating extended attributes on filesystem
             objects, in particular getfattr(1) and setfattr(1).
             An attr(1) command is also provided which is largely compatible
             with the SGI IRIX tool of the same name.
        group:
            utils
        install_date:
            2018-10-24T08:17:51Z
        name:
            attr
        packager:
            Anibal Monsalve Salazar <anibal@debian.org>
        source:
            attr
        url:
            http://savannah.nongnu.org/projects/attr/
        version:
            1:2.4.47-2+b2
```

against

```
uyuni-test:/usr/share/susemanager/salt/packages # salt 'uyuni-test-client.domain' -l error sumautil.debian_info_installed attr
uyuni-test-client.swr.ard:
    ----------
    attr:
        ----------
        arch:
            amd64-deb
        description:
            Utilities for manipulating filesystem extended attributes
             A set of tools for manipulating extended attributes on filesystem
             objects, in particular getfattr(1) and setfattr(1).
             An attr(1) command is also provided which is largely compatible
             with the SGI IRIX tool of the same name.
        epoch:
            1
        group:
            utils
        install_date:
            2018-10-24T08:17:51Z
        install_date_time_t:
            1540369071
        name:
            attr
        packager:
            Anibal Monsalve Salazar <anibal@debian.org>
        release:
            2+b2
        source:
            attr
        url:
            http://savannah.nongnu.org/projects/attr/
        version:
            2.4.47
```

rpasche commented:
Learned a bit more about Salt and made some more improvements. An additional module gets deployed to the Salt minions, especially for Debian systems that override the ```pkg.info_installed``` function so the necessary attributes get returned to the Uyuni server.
I'm also planning to add the apt ```method``` that should be copied to the Debian minions. Also possibly present ```apt-transport-spacewalk``` package has to be removed as this would conflict with the new method. We'll see....

Any feedback is appreciated.

isbm commented:
@rpasche your original commit has been merged to Salt community, regarding getting timestamp. I am also working on the follow-up so we also have `install_date_time_t` part. The other attributes: some of them are likely are going to be a placeholders for backward compatibility.

mcalmer commented on susemanager-utils/susemanager-sls/salt/channels/init.sls:
```diff
@@ -18,12 +18,36 @@ mgrchannels_susemanagerplugin_conf:
+    - name: /etc/susemanager.conf
+    - source:
+      - salt://channels/debian_channel_tokens.repo
+    - user: root
```
AFAIK you use jinja templating in this file. So you need to set it in file.managed. Otherwise it wil just 1:1 deploy the file as it is.
```
file.managed:
  - template: jinja
...
```

isbm commented:
@rpasche this is the PR I was talking about: https://github.com/saltstack/salt/pull/50453

rpasche commented:
I'm still testing the whole process and I also think I'm going to create a new PR when I'm finished with all and I can present cleaner code commits ;-) Also did find some more problems on Debian side. I hope I can test within the next days.

isbm commented:
> Also did find some more problems on Debian side. I hope I can test within the next days.

If they are at Salt side, could we please coordinate that? Otherwise we will have clashes.

rpasche commented:
A further "problem" is the ```architecture```. The ```architecutre``` returned from a x86_64 Debian system is ```amd64```. But within Uyuni (and Spacewalk), x86_64 Debian systems have an architecture name of ```amd64-deb```. And I don't think that it's a good idea to change the architecture within the Salt modules.

So I decided within my tests to write a small module that overrides the ```pkg.info_installed``` method, where I changed the architecture name and also split the debian ```version``` into ```epoch```, ```version``` and ```release``` attributes to match what Uyuni expects. So if the system is not registered to a Uyuni server, you'll get the default values and if the client ```is``` registered to Uyuni, it would use the modified method.

Another "problem" I had during my tests was with disabling (removing) all old repositories (while bootstrapping) with the ```pkg.del_repo``` method. This method updates the apt cache after the repo has been removed. But if you have another repo, that is broken (for whatever reason), that method freezes (you'll see a freezing ```apt-get -q update``` call. So I would suggest to add an attribute ```refresh_db``` or ```force_refresh``` which should default to ```True``` to update the apt cache. But while bootstrapping a Debian client, I would set this to ```False```

One more method is the ```pkg.list_repos``` method. On a Debian system, this returnes a ```dict``` with a ```list``` of repos for each ```source.uri``` as key. Within ```remove_traditional_stack.sls```, this resulted in "Found duplicate ID in 'disable_repo_source.uri'" (sorry, I don't have the error message right now). Currently, I also modified that method on my Debian client system, but I think I will change the ```sls``` to get this fixed.

mcalmer commented:
> A further "problem" is the architecture. The architecutre returned from a x86_64 Debian system is
> amd64. But within Uyuni (and Spacewalk), x86_64 Debian systems have an architecture name of 
> amd64-deb. And I don't think that it's a good idea to change the architecture within the Salt modules.

@rpasche this is typical in original spacewalk code. AFAIK the arch name change happens in the java code. salt should report the original value. I think this is something for @cbbayburt to discuss with you 
and find the best way and place to do the conversion.

rpasche commented:
> 
> 
> > A further "problem" is the architecture. The architecutre returned from a x86_64 Debian system is
> > amd64. But within Uyuni (and Spacewalk), x86_64 Debian systems have an architecture name of
> > amd64-deb. And I don't think that it's a good idea to change the architecture within the Salt modules.
> 
> @rpasche this is typical in original spacewalk code. AFAIK the arch name change happens in the java code. salt should report the original value. I think this is something for @cbbayburt to discuss with you
> and find the best way and place to do the conversion.

@mcalmer Ahh....good to know. I will try that. Thanks.

rpasche commented:
@mcalmer @cbbayburt In my tests, I still have a problem, when I do NOT modify the ```architecture```. If I return the original ```architecture``` (here, ```amd64``` for a x86_64 Debian system), I won't get package update suggestions, as the system will be "up-to-date" within Uyuni (also the software packages links are broken). But if I modify the ```architecture``` (and append ```-deb```, so it returns ```amd64-deb```) to the original architecture, I see possible software packages updates and the software package links are correct. That's why I provided a small ```module```, that overloads the ```pkg.info_installed``` method, so the Debian system should only return the modified ```architecture```, if it has been registered with a Uyuni server.

mateiw commented:
First of all thank you for your contribution @rpasche. We're currently also developing support for Ubuntu/Debian. The work happens in the `ubuntu-support` branch. For the moment the bulk of the work is in PR https://github.com/uyuni-project/uyuni/pull/309. It contains some work similar to yours but with a different approach. 
Feel free to take a look and join our effort.

mbologna commented:
Hi @rpasche !
First of all, thanks for your valuable contribution.
We proceeded in implementing Ubuntu support: you can find implementation details in this PR https://github.com/uyuni-project/uyuni/pull/461, which is already merged.

Updated installer packages are in progress, but as soon as they are ready (take a look at the Uyuni mailing list) we invite you to test them out and let us know if the Ubuntu support is working for you as well.
